telemetry/otel.go

```diff
@@ -0,0 +1,234 @@
+// with default address set to localhost:4317.
+//
+// For testing before setting up and exporting to a collector, use
+// telemetry.StartWithStdoutExport instead.
+//
+// Much of this code is inspired from https://opentelemetry.io/docs/languages/go/getting-started/#initialize-the-opentelemetry-sdk
+package telemetry
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/contrib/detectors/gcp"
+	"go.opentelemetry.io/contrib/instrumentation/runtime"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/stdout/stdoutmetric"
+	"go.opentelemetry.io/otel/exporters/stdout/stdouttrace"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/metric"
+	"go.opentelemetry.io/otel/sdk/metric/metricdata"
+	"go.opentelemetry.io/otel/sdk/resource"
+	"go.opentelemetry.io/otel/sdk/trace"
+)
+
+var defaultStartOnce sync.Once
+
+const (
+	defaultOtelAgentAddr = "0.0.0.0:4317"
+)
+
+type (
+	meterProviderFunc  func(context.Context, *resource.Resource) (*metric.MeterProvider, error)
+	tracerProviderFunc func(context.Context, *resource.Resource) (*trace.TracerProvider, error)
+)
+
+// Start initializes automatic instrumentation for exporting OTLP metrics and traces
+// to an OpenTelemetry collector agent.
+func Start(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newOTLPTraceProvider, newOTLPMeterProvider)
+	})
+	return
+}
+
+// StartWithStdoutExport initializes automatic instrumentation for exporting OTLP
+// metrics and traces to stdout (e.g. Cloud Logging if using Cloud Run).
+//
+// This is intended for testing and debugging metrics and traces before exporting
+// them to an OpenTelemetry collector agent.
+func StartWithStdoutExport(ctx context.Context) (shutdown func(context.Context) error, err error) {
+	defaultStartOnce.Do(func() {
+		shutdown, err = start(ctx, newStdoutTraceProvider, newStdoutMeterProvider)
+	})
+	return
+}
+
+func start(ctx context.Context, newTracer tracerProviderFunc, newMeter meterProviderFunc) (shutdown func(context.Context) error, err error) {
+	var shutdownFuncs []func(context.Context) error
+
+	shutdown = func(ctx context.Context) error {
+		var err error
+		for _, fn := range shutdownFuncs {
+			err = errors.Join(err, fn(ctx))
+		}
+		shutdownFuncs = nil
+		return err
+	}
+
+	if err := runtime.Start(); err != nil {
+		return nil, fmt.Errorf("failed to start runtime instrumentation: %w", err)
+	}
+	// Set up some automatic attributes regarding deployment environment, service,
+	// GCP, etc to attach to telemetry.
+	res, err := resource.New(
+		ctx,
+		resource.WithFromEnv(),
+		resource.WithProcess(),
+		resource.WithDetectors(gcp.NewDetector()),
```

Comment:
What happens if this is run outside of GCP?

Reply:
This library is scoped to Cloud Run and GKE applications.

But it should not be a problem as the resource detection processor just fetches any GCP-related semantic attributes set by the runtime.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/processor/resourcedetectionprocessor/README.md#gcp-metadata